### PR TITLE
fix(server): close silent message-drop paths behind reconnect + XEP-0198

### PR DESCRIPTION
**Pull request by randax**
## Summary

Follow-up to #160 / #163 / #164. Users still reported the browser chat reconnecting often, messages dropping across reconnects, and the Send button freezing while recovery ran. A prod log audit on `sha-9bfaac4` surfaced two silent-drop paths the previous PRs didn't address plus three client-side regressions that together produce those symptoms. Each fix is one commit; none change XEP-0198 wire behaviour.

### Server-side

- **`fix(server): surface silent broadcast drops via typed BroadcastOutcome`** — `ConnectionRegistry::try_send_to` returned `bool` and `debug!`'d Full-channel drops. The groupchat broadcast loop ignored the return value and logged `recipients = occupants.len()` (the *intended* count, not the *delivered* one). A backpressured 256-slot outbound mpsc fills in ~2s under even mild load and then drops stanzas with zero signal. Fixed by returning a typed `BroadcastOutcome { Delivered, NotConnected, DroppedFull, DroppedClosed }`, `warn!`-ing Full, adding four Prometheus counters (`waddle_broadcast_{delivered,not_connected,dropped_full,dropped_closed}_total`), and changing the broadcast log to emit `intended/delivered/dropped_full/dropped_closed/not_connected`.
- **`fix(server): surface XEP-0198 unacked-queue eviction via UnackedPushResult`** — `UnackedQueue::push` silently `pop_front`'d the oldest stanza when the 1000-entry cap was hit. A later `<resumed/>` for the same session replayed a buffer missing the earliest (most load-bearing) stanzas with zero signal — the sender-side half of "reconnects drop messages". Fixed by returning `UnackedPushResult { Accepted, Evicted(UnackedStanza) }`, marking `push` `#[must_use]`, and having `StreamManagementState::record_outbound` `warn!` + bump `waddle_sm_unacked_evicted_total`. Eviction policy itself is unchanged — the follow-up decision of eviction vs stream-close is easier to make with the counter in place.
- **`fix(server): parse SM countable stanzas by element name, not prefix`** — `is_countable_stanza` used `starts_with("<message")` / `"<iq"` / `"<presence"`, which is the same substring-matching pattern that bit PR #164. Replaced with `Element::from_str` + `matches!(element.name(), …)`. All existing call sites serialize via minidom so every frame has a namespace — no wire change.

### Client-side

- **`fix(chat): remove persisted queue entries on SM ack, not on send return`** — `flushQueuedRoomMessages` / `flushQueuedDirectMessages` called `removeQueuedMessage` the moment `messaging.sendGroupMessage` returned a non-null id. But that return is just the stanza-js-assigned UUID, not evidence of wire write — stanza.js can buffer the frame against a dead transport and the server can silently drop it per the broadcast-channel Full path above. Fixed by deferring removal to the `message:acked` handler (XEP-0198) and tracking an in-memory `inflightQueuedIds` set so a second flush in the same session is a no-op for already-handed-off entries. Fresh sessions clear the set because stanza.js dropped its SM buffer there. `message:failed` clears only the in-memory set so the next flush re-sends.
- **`fix(chat): stop blocking Send on reconnect + room switch`** — `sendGroupMessage` awaited `requireJoinedRoom` → `switchRoom` (10s self-presence timeout) → `connect` (15s reconnect timeout), which is how the Send button could freeze for ~25s on a bad network. `sendDirectMessage` had the same shape. Both send paths now take a fast path on the ready state (`roomIsReady(roomJid)` / `canUseConnectedSession()`) and enqueue optimistically otherwise, driving recovery in the background. `waitForRoomSelfPresence` dropped from 10s → 4s; the `muc:available` stream papers over single-request lag on the next presence. The dead `isLocalAvailabilityError` helper is removed.

### Live evidence that drove the fixes

Prod pod `waddle-server-7d8d67bbc-8v5rt` (image `sha-9bfaac4`, all three prior PRs deployed) shows ~125 groupchat broadcasts/sec sustained from one user to a room with `recipients=1`. With 256-slot outbound channels and 1000-entry SM queues, every recipient in a loaded room is a candidate for silent drops on both paths. The new counters make future incidents self-diagnosing instead of requiring a code-read.

Full write-up of the audit is in `/Users/oyr/.claude/plans/i-can-still-see-pure-aurora.md` (local plan file, not in the repo).

## Test plan

- [x] \`cargo test -p waddle-xmpp -p waddle-server\` — 213 + 3 pass (one flaky port-file integration test passes on retry; unrelated)
- [x] \`cargo fmt --check\` — clean
- [x] \`cargo clippy -p waddle-xmpp -p waddle-server --all-targets\` — clean on touched code (pre-existing \`handle_iq\` / \`insert_channel\` / \`too_many_arguments\` warnings are not caused by this PR, same as #164)
- [x] \`bun test\` in \`chat/\` — 236 pass
- [x] \`bun run lint\` (knip) in \`chat/\` — clean
- [ ] Deploy to prod; watch new counters in Grafana:
  - \`waddle_broadcast_dropped_full_total\` should be near-zero on a healthy day. Any non-zero rate is a real silent drop we used to ship.
  - \`waddle_sm_unacked_evicted_total\` likewise.
- [ ] Devtools reproduction for fix #3: send a message → throttle to Offline → wait 10s → Online. Confirm the \`waddle.chat.outbound-queue.*\` entry in localStorage persists until the eventual \`message:acked\` (not the synchronous send return), and that \`message:failed\` keeps the entry queued for the next reconnect.
- [ ] Devtools reproduction for fix #4: with the tab backgrounded, click Send. The button must react in <500ms and the status telegraph read "Message queued until the connection returns", not a spinner on the Send itself.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: server/crates/waddle-server/src/server/routes/websocket.rs [hunk 1]
```diff
@@ -2779,14 +2793,23 @@ async fn handle_message(
...
+                    BroadcastOutcome::NotConnected => not_connected += 1,
+                }
             }
         }
 
         info!(
             room = %room_jid,
             sender = %sender_nick,
-            recipients = occupants.len(),
+            intended = occupants.len(),
```
[the remark this answers is not in the document]
Comment (randax) [reply server/crates/waddle-server/src/server/routes/websocket.rs:2829]: Fixed in 9ed0e20: sender echo now counts as `delivered`, and `debug_assert_eq!` pins the invariant `occupants.len() == delivered + dropped_full + dropped_closed + not_connected`. A 1-person room now logs `intended=1 delivered=1`. Thanks for catching this.

File: server/crates/waddle-server/src/server/routes/websocket.rs [hunk 2]
```diff
@@ -2779,14 +2793,23 @@ async fn handle_message(
...
+            delivered,
+            dropped_full,
+            dropped_closed,
+            not_connected,
             "Groupchat message broadcast"
```
[the remark this answers is not in the document]
Comment (randax) [reply server/crates/waddle-server/src/server/routes/websocket.rs:2834]: Same issue as the Codex comment below — fixed in 9ed0e20. Sender echo now increments `delivered`, so the accounting line adds up.

File: server/crates/waddle-server/src/server/routes/websocket.rs [hunk 3]
```diff
@@ -634,11 +634,17 @@ fn parse_sasl_auth_frame(frame: &str) -> Result<(String, String), String> {
...
+/// name and never build a surprise in future.
 fn is_countable_stanza(frame: &str) -> bool {
-    let trimmed = frame.trim_start();
-    trimmed.starts_with("<iq")
-        || trimmed.starts_with("<message")
-        || trimmed.starts_with("<presence")
+    let Ok(element) = Element::from_str(frame.trim_start()) else {
+        return false;
+    };
+    matches!(element.name(), "iq" | "message" | "presence")
```
[the remark this answers is not in the document]
Comment (randax) [reply server/crates/waddle-server/src/server/routes/websocket.rs]: Fixed in 9ed0e20: replaced `Element::from_str` with a byte-level scan — `trim_start` → `strip_prefix('<')` → `find(whitespace/`>`/`/`)` → `matches!` on the slice. Same correctness (no substring collisions), no per-stanza DOM allocation on the SM hot path. Existing `is_countable_stanza_matches_element_name_not_prefix` test continues to cover every edge case.

File: server/crates/waddle-xmpp/src/registry/connection_registry.rs [hunk 4]
```diff
@@ -252,23 +284,37 @@ impl ConnectionRegistry {
...
+            }
             Err(tokio::sync::mpsc::error::TrySendError::Full(_)) => {
-                debug!(jid = %jid, "Outbound channel full; dropping broadcast stanza");
-                false
+                prometheus::increment_broadcast_dropped_full();
+                warn!(
+                    jid = %jid,
+                    "Outbound channel full; broadcast stanza dropped silently"
+                );
+                BroadcastOutcome::DroppedFull
```
[the remark this answers is not in the document]
Comment (randax) [reply server/crates/waddle-xmpp/src/registry/connection_registry.rs:319]: Fixed in 54e74a4: demoted to `debug!`. The aggregated "Groupchat message broadcast" info log already carries a per-send `dropped_full` total and `waddle_broadcast_dropped_full_total` is always-on in Prometheus, so the per-recipient line was pure triage detail.

File: chat/src/lib/xmpp/client.ts [hunk 5]
```diff
@@ -1133,17 +1138,35 @@ export class BrowserXmppClient {
...
+    // landed on the server. Removing the entry here instead of on the
+    // synchronous `sendMessage` return is what closes the "stanza.js
+    // buffered it but the wire write never went through" window that was
+    // wiping messages out of localStorage before they were durable.
     xmpp.on("message:acked", (msg) => {
       if (this.xmpp !== xmpp) return;
-      if (msg?.id) this.messageAckHandler?.(msg.id);
+      if (msg?.id) {
+        this.inflightQueuedIds.delete(msg.id);
+        removeQueuedMessage(this.queueScope, msg.id);
```
[the remark this answers is not in the document]
Comment (randax) [reply chat/src/lib/xmpp/client.ts]: Fixed in 2ff45ca, using your suggestion verbatim: `if (this.inflightQueuedIds.delete(msg.id)) removeQueuedMessage(...)`. Live-send acks now skip the localStorage read+write entirely.
